walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,146 @@
 
 package com.google.sps;
 
+import java.util.Collections;
 import java.util.Collection;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    // This variable will hold all of the invalid meeting times.
+    // This list is sorted by start times, and has no overlapping times.
+    List<TimeRange> invalidTimeRanges = new ArrayList<>();
+    
+    // This is the list of the attendees which attend the meeting.
+    Collection<String> attendees = request.getAttendees();
+
+    // The duration of the meeting request.
+    int meetingDuration = (int) request.getDuration();
+
+    // Populate invalidTimeRanges variable, and sort by increasing start time.
+    populateInvalidTimeRanges(invalidTimeRanges, events, attendees);
+    Collections.sort(invalidTimeRanges, TimeRange.ORDER_BY_START);
+
+    // Merge conflicting TimeRanges.
+    mergeConflictingTimeRanges(invalidTimeRanges);
+
+    // Get the inverse / valid TimeRanges of the invalidTimeRanges.
+    List<TimeRange> validTimeRanges = getValidTimeRanges(invalidTimeRanges, 
+      meetingDuration);
+
+    return validTimeRanges;
+
+  }
+
+	/**
+   * Returns true if the duration (presumably of a TimeRange) is longer than
+   * the meeting duration request.
+   */
+  private boolean fitsMeetingDuration(int duration, int meetingDuration) {
+    return duration >= meetingDuration; 
+  }
+
+  /**
+   * Get the valid (inverse) TimeRanges from invalidTimeRanges.
```

Comment:
The ordering of invalidTimeRanges is very important for this method to work, you should probably call that out